### PR TITLE
Merge RBAC branch into master

### PR DESCRIPTION
**Pull request by Bento007**
This adds resource based access control to fusillade. Admins can now create a resource type, create resource policies for that resource type, add resource ids under that resource types, and assign users and groups access to resource ids. In order for a user to have access to a resource, they must have a role that grants access either through a group or directly, and the user must be given access to a resource. 

### Test plan
 Adding test cases for all new APIs

### Deployment instructions & migrations
- Upgrade the schema to the latest

### Release notes

- Adding endpoints to create resource types and resource ids
- Adding endpoints to manage resource policies attached to resource types
- Adding endpoints to give users access to controlled resources.

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -241,20 +241,21 @@ For resource, set the `partition` to `hca`, set your `service` name to the name
 A new `resource_type` is created by providing the name of the `resource_type`, and the actions that can be performed on it. 
  Once a `resource_type` is created you can store `resource_id`s of that `resource_type` to apply ACLs.
  
-An `access_policy` is a policy associated with a `resource_type` and is used to define the different access levels 
+An `resource_policy` refers to a policy associated with a `resource_type` and is used to define different access levels
```
Comment (chmreid) [inline Readme.md]:
```suggestion
A `resource_policy` refers to a policy associated with a `resource_type` and is used to define different access levels 
```

File: Readme.md [hunk 2]
```diff
@@ -241,20 +241,21 @@ For resource, set the `partition` to `hca`, set your `service` name to the name
...
-An `access_policy` is a policy associated with a `resource_type` and is used to define the different access levels 
+An `resource_policy` refers to a policy associated with a `resource_type` and is used to define different access levels 
  between principals and `resource_id`s. A principal may have only one level of access to a `resource_id`. All 
- `resource_id`s use the same pool of access policies for that particular `resource_type`. This mean that modifying an `access_policy`, 
- modifies it for all principals with that access level to a `resource_id`. New access policies can be defined for a 
- `resource_type` after the `resource_type` has been created. Deleting an `access_policy` removes access for all principals 
- with that level of access between a `resource_id`.
- Access policies can only define policies that use actions supported by that `resource_type`. 
- Actions can be added and removed after a `resource_type` has been created.
+ `resource_id`s use the same pool of resource policies for that particular `resource_type`. This mean that modifying an
```
Comment (chmreid) [inline Readme.md]:
```suggestion
 `resource_id`s use the same pool of resource policies for that particular `resource_type`. This mean that modifying a
```

File: Readme.md [hunk 3]
```diff
@@ -241,20 +241,21 @@ For resource, set the `partition` to `hca`, set your `service` name to the name
...
+ `resource_policy`, modifies it for all principals with that access level to a `resource_id`. New `resource_policy` can
+  be 
+ defined for a `resource_type` after the `resource_type` has been created. Deleting an `resource_policy` removes access
```
Comment (chmreid) [inline Readme.md]:
```suggestion
 defined for a `resource_type` after the `resource_type` has been created. Deleting a `resource_policy` removes access 
```

**Comment by chmreid** [comment]
To clarify - was the `tsmith/resource-feat` branch the base branch for the sequence of resource-related PRs we've been reviewing?

**Comment by Bento007** [comment]
@chmreid yes

File: fusillade/directory/clouddirectory.py [hunk 4]
```diff
@@ -463,6 +463,25 @@ def _make_attribute(name: str, value: any):
...
+        result = dict()
+        # check if we are parsing object attributes or typed link attributes
+        typed_link = attributes[0].get('Key') is None
+        for attr in attributes:
+            key = attr['AttributeName'] if typed_link else attr['Key']['Name']
+            if ValueTypes.StringValue.name in attr['Value'].keys():
+                result[key] = attr['Value'][ValueTypes.StringValue.name]
+            elif ValueTypes.BinaryValue.name in attr['Value'].keys():
+                result[key] = attr['Value'][ValueTypes.BinaryValue.name]
+            elif ValueTypes.BinaryValue.name in attr['Value'].keys():
```
[the remark this answers is not in the document]
Comment (Bento007) [reply fusillade/directory/clouddirectory.py]: nice catch!

File: fusillade/directory/clouddirectory.py [hunk 5]
```diff
@@ -765,43 +798,43 @@ def get_link_attributes(self, TypedLinkSpecifier, AttributeNames, **kwargs) -> D
...
-        :param policy_type: the type of policies to retrieve from the policy nodes
-        :return: returns the policies of the type IAMPolicy from a list of policy paths.
-        """
+    def get_policy_ids(self, policy_paths: List[Dict[str, Any]], policy_type='IAMPolicy'):
         # Parse the policyIds from the policies path. Only keep the unique ids
-        policy_ids = set(
+        return set(
             [
-                o['PolicyId']
+                f"${o['PolicyId']}"
```
[the remark this answers is not in the document]
Comment (Bento007) [reply fusillade/directory/clouddirectory.py]: yeah, it's a cloud directory thing

**Pull request merged**